packages/server/lib/browsers/bidi_automation.ts

```diff
@@ -0,0 +1,267 @@
+// Taken from https://fetch.spec.whatwg.org/#request-initiator-type
+type RequestInitiatorType = 'audio' | 'beacon' | 'body' | 'css' | 'early-hints' | 'embed' | 'fetch' | 'font' | 'frame' | 'iframe' | 'image' | 'img' | 'input' | 'link' | 'object' | 'ping' | 'script' | 'track' | 'video' | 'xmlhttprequest' | 'other' | null
+// Taken from https://fetch.spec.whatwg.org/#concept-request-destination
+type RequestDestination = 'audio' | 'audioworklet' | 'document' | 'embed' | 'font' | 'frame' | 'iframe' | 'image' | 'json' | 'manifest' | 'object' | 'paintworklet' | 'report' | 'script' | 'serviceworker' | 'sharedworker' | 'style' | 'track' | 'video' | 'webidentity' | 'worker' | 'xslt' | ''
+
+export type NetworkBeforeRequestSentParametersModified = NetworkBeforeRequestSentParameters & {
+  request: {
+    destination: RequestDestination
+    initiatorType: RequestInitiatorType
+  }
+}
+
+// maps the network initiator to a ResourceType (which is initially based on CDP).
+// This provides us with consistency of types in our request/response middleware, which is important for cy.intercept().
+const normalizeResourceType = (type: RequestInitiatorType): ResourceType => {
+  switch (type) {
+    case 'css':
+      return 'stylesheet'
+    case 'xmlhttprequest':
+      return 'xhr'
+    case 'img':
+      return 'image'
+    case 'iframe':
+      return 'document'
+      // for types we cannot determine, we can set to other.
+    case 'audio':
+    case 'beacon':
+    case 'body':
+    case 'early-hints':
+    case 'embed':
+    case 'frame':
+    case 'input':
+    case 'link':
+    case 'object':
+    case 'track':
+    case 'video':
+    case null:
+      return 'other'
+    default:
+      return type
+  }
+}
+
+export class BidiAutomation {
+  // events needed to subscribe to in order for our BiDi automation to work properly
+  static BIDI_EVENTS = [
+    'network.beforeRequestSent',
+    'network.responseStarted',
+    'network.responseCompleted',
+    'network.fetchError',
+    'browsingContext.contextCreated',
+    'browsingContext.contextDestroyed',
+  ]
+  private webDriverClient: WebDriverClient
+  private automation: Automation
+  private autContextId: string | undefined = undefined
+  // set in firefox-utils when creating the webdriver session initially and in the 'reset:browser:tabs:for:next:spec' automation hook for subsequent tests when the top level context is recreated
+  private topLevelContextId: string | undefined = undefined
+  private interceptId: string | undefined = undefined
+
+  private constructor (webDriverClient: WebDriverClient, automation: Automation) {
+    this.automation = automation
+    this.webDriverClient = webDriverClient
+
+    // bind Bidi Events to update the standard automation client
+    // Error here is expected until webdriver adds initiatorType and destination to the request object
+    // @ts-expect-error
+    this.webDriverClient.on('network.beforeRequestSent', this.onBeforeRequestSent.bind(this))
```

Comment:
declare these with an arrow function to get the correct this context. As of right now, these are unique references that are NOT getting cleaned up